raiden/api/python.py

```diff
@@ -137,47 +141,62 @@ def channel_open(
             settle_timeout=None,
             reveal_timeout=None,
             poll_timeout=DEFAULT_POLL_TIMEOUT):
-        """ Open a channel with the peer at `partner_address`
-        with the given `token_address`.
-        """
-        if reveal_timeout is None:
-            reveal_timeout = self.raiden.config['reveal_timeout']
 
-        if settle_timeout is None:
-            settle_timeout = self.raiden.config['settle_timeout']
+        # Prevent concurrent attempts to open a channel with the same token and
+        # partner address
+        if partner_address not in self.open_channel_locks:
+            new_open_lock = gevent.lock.Semaphore()
+            self.open_channel_locks[(token_address, partner_address)] = new_open_lock
+
+        open_lock = self.open_channel_locks[(token_address, partner_address)]
 
-        if settle_timeout <= reveal_timeout:
-            raise InvalidSettleTimeout(
-                'reveal_timeout can not be larger-or-equal to settle_timeout'
+        if open_lock.locked():
+            raise DuplicatedChannelError(
```

Comment:
This exception is being raised too early. This exception should be

> Raised if someone tries to create a channel that already exists.

At this point, the transaction may still be going, which may be surprising to the caller (because the channel does not exist yet). A better approach would be (pseudo-code):

```python
if not is_new_channel():
  return

lock = get_or_create_lock()
if lock.acquire(blocking=False):
   # only run one open transaction
   self._open_channel()
   lock.release()
else:
  # if another thread sent the transaction, wait for it to finish
  lock.wait()
```

Note: The above code should go in the `ChannelManager`